Replace Arg.Bad with fatal in compenv.ml

bschommer opened the pull request:
Since there is no exception handler for Arg.Bad the exception is uncaught.
Now the compiler aborts with fatal and prints out the error message.

This should fix [MPR#7460](https://caml.inria.fr/mantis/view.php?id=7460).

bschommer committed:
Replace Arg.Bad with fatal in compenv.ml

Since there is no exception handler for Arg.Bad the exception is uncaught.
Now the compiler aborts with fatal and prints out the error message.

bschommer commented:
I'm not sure if a Change entry is need and in my opinion the error message can be improved a bit but I couldn't come up with a better one at the moment.

gasche commented:
I think that, if someone reported a bug for it, it deserves a Changes entry.

Could you maybe add a regression test among the `tool-*` tests, possibly in a new directory `tool-command-line`?

bschommer commented:
No problem.

bschommer committed:
Added Chang entry.

bschommer committed:
Improved error message.

bschommer committed:
Added test and reference.

Octachron commented:
Since this is an argument parsing error, I would argue that it makes sense to restore 4.03 behavior and fail with a standard Arg error message (with the addition of the usage message).

bschommer committed:
Update reference.

bschommer commented:
I personally find the old error message not that usefull, however I'm not in strong favor of my new message and if it is wanted I can revert to the old message.

Printing the help and usage message would require either changing the interface of `compenv.ml` or a lot of code duplication in each of the tools that use the `Arg` module and `compilenv.ml`.

gasche commented:
I think that the wording of the new error message is not very good. The older one is maybe a bit better -- it's not terribly useful, as you say, but it reads better as a sentence.

bschommer commented:
I had a look and the function that raises the Exception `Arg.Bad` is only called in the bytecode and the native compiler. I will rename the warning and restore the behavior of 4.03.

bschommer committed:
Changed implementation to 4.03 behavior.

The function raising Arg.Bad is only used in driver/mail.ml and
driver/optmain.ml and so adding the two exception handlers restored the
4.03 behavior.

bschommer committed:
Fix linking issue.

gasche commented:
I don't understand the last change. I noticed with the CI that there was an issue with the patch, but what is the issue? Why is 7cf0cfac08fddf7beb284cedfbbf46f07e269de5 useful? I'm not so happy with the fact that you are changing the `Compenv` interface. I suppose you are trying to avoid some module dependencies, it's not `Compenv->Arg` (this dependency was already present), maybe it is `Compenv->Options`? Why is this dependency bad?

bschommer commented:
I actually have no idea, since every function I used in `main.ml` and `optmain.ml` was used before or came from a the `Arg` module to which there was already a dependency.

gasche commented:
Could we find a solution, then, that does not change the `Compilenv` interface? One may need to update the `.depend` file to reflect a new dependency from `Compilenv` on `Options`. If that does not work, we might drop the idea of showing the usage text in this case. But I would like to understand what the linking issue was (why would `Dynlinkaux` be affected?) in case it would reflect a bug in this patch (sounds unlikely) or another recent change.

Octachron commented:
Since `Compenv.process_deferred_actions` is only called from `main.ml` and `optmain.ml`, I think that an alternative to the current patch would be to catch the `Arg.Bad` exception in these two modules without modifying `Compenv`, except  for documenting the fact that Compenv.process_deferred_actions can raise an Arg.Bad exception.

bschommer commented:
I actually tried this in commit eee2829d5405d8e1ae10cf46ef12f02a0cab8349 but I got some weird linking errors.

bschommer committed:
Revert "Fix linking issue."

This reverts commit 7cf0cfac08fddf7beb284cedfbbf46f07e269de5.

bschommer committed:
Revereted changes to interface.

The problem was the missing begin...end block plus the missing exit 2.

bschommer commented:
I think I found the source of my link problems (beside the missing `begin ... end` block) was the missing `exit 2`. However I'm still curious why this generated such strange link errors. The test need to be adopted a bit in order to reflect the different option set of flambda and non-flambda ocamlopt.

bschommer committed:
Fixed test and removed whitespace commit.

bschommer committed:
Upated compend.ml

Removed whitespace

bschommer commented:
It should work now as 4.03 and before. The test filters out all options beside the message for the unknown file, this should avoid breaking the test whenever a new option is introduced.

gasche commented:
You could maybe test only ocamlc?

gasche commented on Changes:
```diff
@@ -37,6 +37,10 @@ Next version (4.05.0):
 
+- PR#7460, GPR#1011: catch exception thrown when unknown files are passed
+  as argument.
+  (Bernhard Schommer)
```
Maybe point out that @Octachron was helpful in reviewing? (Or maybe you had found the nasty syntax thing independently?) Otherwise it now seems good to merge.

bschommer replied:
No problem, will amend the change entry.

bschommer committed:
Amend Change entry with people involved.

gasche commented:
Thanks to all, I merged in trunk and 4.04. (I was a bit hesitant to merge in 4.04 as this is admittedly a very minor regression, but the final patch is also very easy to check for non-intrusiveness so I went for it.)

gasche commented:
In a48b2cb03b8e3b001668b052224269a37a21948b I updated the new test to work also on bytecode-only configurations. (I also forget about those when I write tests.)
